File: service/internal/model/identity.go
package model

import (
	"time"

	"github.com/google/uuid"
	"github.com/rs/zerolog/log"
)

// Identity represents the identity of some model resource.
type Identity struct {
	ID      uuid.UUID
	Version uuid.UUID
	Created time.Time
	Updated time.Time
}

// Create a new Identity object for a new model instance.
func NewIdentity() (Identity, error) {
	id, err := uuid.NewRandom()
	if err != nil {
		log.Warn().Err(err).Msg("Failed to generate ID")
		return Identity{}, err
	}

	version, err := uuid.NewRandom()
	if err != nil {
		log.Warn().Err(err).Msg("Failed to generate Version")
		return Identity{}, err
	}

	return Identity{
		ID:      id,
		Version: version,
		Created: time.Now(),
		Updated: time.Now(),
	}, nil
}

File: service/crates/model/src/identity.rs
use chrono::{DateTime, Utc};
use uuid::Uuid;

/// The identity of some model resource
#[derive(Debug, PartialEq)]
pub struct Identity<I> {
    /// The actual ID of the resource
    pub id: I,
    /// A version tag of the resource
    pub version: Uuid,
    /// When the resource was created
    pub created: DateTime<Utc>,
    /// When the resource was last updated
    pub updated: DateTime<Utc>,
}

impl<I> Identity<I> {
    /// Create a new Identity instance for the given ID value to represent a brand new resource
    ///
    /// # Parameters
    /// - `id` - The ID of the new resource
    ///
    /// # Returns
    /// The Identity object
    pub fn new(id: I) -> Self {
        let now = Utc::now();

        Self {
            id,
            version: Uuid::new_v4(),
            created: now,
            updated: now,
        }
    }
}

impl<I> Default for Identity<I>
where
    I: Default,
{
    /// Create a new Identity instance with a default ID value to represent a brand new resource
    ///
    /// # Returns
    /// The Identity object
    fn default() -> Self {
        Self::new(I::default())
    }
}
